### PR TITLE
Rework user's PR for adding servlet attributes access

### PR DESCRIPTION
**Pull request by cy6erGn0m**
**Subsystem**
ktor-server-servlet, ktor-server-jetty

**Motivation**
[KTOR-424 Get client certificate information from request](https://youtrack.jetbrains.com/issue/KTOR-424)

**Solution**
Rework user's [PR #1771](https://github.com/ktorio/ktor/pull/1771) to provide access to servlet attributes. In particular, it provides a way to look at client' certificate from servlet request attributes.

### CODE REVIEW

File: ktor-server/ktor-server-jetty/jvm/test/io/ktor/tests/server/jetty/JettyEngineTest.kt [hunk 1]
```diff
@@ -4,15 +4,62 @@
...
+                            request: HttpServletRequest?,
+                            response: HttpServletResponse?
+                        ) {
+                            request?.setAttribute("ktor.test.attribute", "135")
+                            delegate?.handle(target, baseRequest, request, response)
+                        }
+                    }
+                }
+            })
+
```
Comment (e5l) [inline ktor-server/ktor-server-jetty/jvm/test/io/ktor/tests/server/jetty/JettyEngineTest.kt:46]: empty line

File: ktor-server/ktor-server-jetty/jvm/test/io/ktor/tests/server/jetty/JettyEngineTest.kt [hunk 2]
```diff
@@ -4,15 +4,62 @@
...
     CompressionTestSuite<JettyApplicationEngine, JettyApplicationEngineBase.Configuration>(Jetty)
 
 class JettyContentTest : ContentTestSuite<JettyApplicationEngine, JettyApplicationEngineBase.Configuration>(Jetty)
 
-class JettyHttpServerTest : HttpServerTestSuite<JettyApplicationEngine, JettyApplicationEngineBase.Configuration>(Jetty)
+class JettyHttpServerTest : HttpServerTestSuite<JettyApplicationEngine, JettyApplicationEngineBase.Configuration>(Jetty) {
+    override fun configure(configuration: JettyApplicationEngineBase.Configuration) {
+        super.configure(configuration)
+        configuration.configureServer = {
+            addLifeCycleListener(object : AbstractLifeCycle.AbstractLifeCycleListener() {
```
Comment (e5l) [inline ktor-server/ktor-server-jetty/jvm/test/io/ktor/tests/server/jetty/JettyEngineTest.kt:29]: consider extracting block to improve readability

File: ktor-server/ktor-server-servlet/jvm/src/io/ktor/server/servlet/Attributes.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import io.ktor.application.*
+import io.ktor.server.engine.*
+import io.ktor.util.*
+import javax.servlet.*
+
+/**
+ * Provides javax.servlet request attributes or fail it the underlying engine is not
+ * servlet-backed.
+ */
+public val ApplicationCall.servletRequestAttributes: Map<String, Any>
```
Comment (e5l) [inline ktor-server/ktor-server-servlet/jvm/src/io/ktor/server/servlet/Attributes.kt:16]: Should we put it to the request directly?  like `call.request.servletAttributes`

File: ktor-server/ktor-server-servlet/jvm/src/io/ktor/server/servlet/AsyncServlet.kt [hunk 4]
```diff
@@ -42,6 +42,10 @@ public open class AsyncServletApplicationCall(
             putResponseAttribute(it)
         }
     }
+
+    init {
+        servletRequest.putServletAttributesTo(this)
```
Comment (e5l) [inline ktor-server/ktor-server-servlet/jvm/src/io/ktor/server/servlet/AsyncServlet.kt:47]: It's hard to read such API, please consider to swap receiver and argument. `addServletAttributes(servletRequest)`

**Review by e5l — APPROVED** [review]
LGTM in general

File: ktor-server/ktor-server-servlet/jvm/src/io/ktor/server/servlet/Attributes.kt (hunk 3 above)
Comment (cy6erGn0m) [reply ktor-server/ktor-server-servlet/jvm/src/io/ktor/server/servlet/Attributes.kt:16]: Yes, it's possible. Actually, it is what the original PR proposes.

**Comment by cy6erGn0m** [comment]
Merged

**Pull request closed without merging**